web/src/features/fireWeather/components/graphs/PrecipGraph.tsx

```diff
@@ -0,0 +1,364 @@
+    const datesFromAllSources: Date[] = []
+
+    const aggreObservedPrecips: { [k: string]: number } = {}
+    observedValues.forEach(({ datetime, precipitation: precip }) => {
+      const date = formatDateInPDT(datetime, 'YYYY-MM-DD')
+      if (!aggreObservedPrecips[date]) {
+        aggreObservedPrecips[date] = Number(precip)
+      } else {
+        aggreObservedPrecips[date] = aggreObservedPrecips[date] + Number(precip)
+      }
+    })
+
+    const _observedPrecips = Object.entries(aggreObservedPrecips).map(
+      ([formattedDate, totalPrecip]) => {
+        const date = moment(formattedDate)
+          .utc()
+          .set({ hour: Math.abs(utcOffset), minute: 0 })
+          .toDate()
+        datesFromAllSources.push(date)
+
+        return {
+          date,
+          value: Number(totalPrecip.toFixed(2))
+        }
+      }
+    )
+
+    const _forecastPrecips = forecastValues.map(({ datetime, total_precipitation }) => {
+      const date = moment(datetime)
+        .utc()
+        .set({ hour: Math.abs(utcOffset) })
+        .toDate()
+      datesFromAllSources.push(date)
+
+      return {
+        date,
+        value: Number(total_precipitation.toFixed(2))
+      }
+    })
+
+    const currDate = new Date()
+    const pastDate = moment(currDate)
+      .subtract(5, 'days')
+      .toDate()
+    const [minDate, maxDate] = d3.extent(datesFromAllSources)
+    let d1 = minDate || pastDate
+    let d2 = maxDate || currDate
+    d1 = moment(d1)
+      .subtract(6, 'hours')
+      .toDate()
+    d2 = moment(d2)
+      .add(6, 'hours')
+      .toDate()
+    const _xDomain: [Date, Date] = [d1, d2]
+
+    return {
+      xDomain: _xDomain,
+      xTickValues: d3Utils.getTickValues(_xDomain, utcOffset, false),
+      observedPrecips: _observedPrecips,
+      forecastPrecips: _forecastPrecips
+    }
+  }, [utcOffset, observedValues, forecastValues])
+
+  // Effect hook for displaying graphics
+  useEffect(() => {
+    if (svgRef.current) {
+      /* Clear previous graphics before rendering new ones */
+      d3.select(svgRef.current)
+        .selectAll('*')
+        .remove()
+
+      const svg = d3
+        .select(svgRef.current)
+        .attr('viewBox', `0 0 ${svgWidth} ${svgHeight}`)
+
+      const chart = svg
+        .append('g')
+        .attr('class', 'chart')
+        .attr('transform', `translate(${margin.left}, ${margin.top})`)
+
+      const context = svg // svg group for Y axis and its labels
+        .append('g')
+        .attr('class', 'context')
+        .attr('transform', `translate(${margin.left}, ${margin.top})`)
+
+      const legendMarginTop = chartHeight + 40
+      const legend = svg
+        .append('g')
+        .attr('class', 'legend')
+        .attr('transform', `translate(${margin.left}, ${legendMarginTop})`)
+
+      /* Create scales for x and y axes */
+      const xScale = d3
+        .scaleTime()
+        .domain(xDomain)
+        .range([0, chartWidth])
+      const xScaleCopy = xScale.copy()
```

Comment:
what's the purpose of making `xScaleCopy` rather than just using `xScale`?

Reply:
This might be better to talk about it over a chat